libsplinter/src/service/scabbard/state.rs

```diff
@@ -342,42 +342,64 @@ impl ScabbardState {
 
 fn receipts_into_transact_state_changes(
     receipts: &[TransactionReceipt],
-) -> Vec<TransactStateChange> {
-    receipts
+) -> Result<Vec<TransactStateChange>, ScabbardStateError> {
+    if receipts.iter().any(is_receipt_invalid) {
+        return Err(ScabbardStateError(
+            "invalid transaction receipts cannot be converted into state changes".into(),
+        ));
+    }
+
+    Ok(receipts
         .iter()
-        .flat_map(|receipt| {
-            receipt
-                .state_changes
-                .iter()
-                .cloned()
-                .map(|change| match change {
+        .flat_map(|receipt| match &receipt.transaction_result {
+            TransactionResult::Valid { state_changes, .. } => {
+                state_changes.iter().cloned().map(|change| match change {
                     transact::protocol::receipt::StateChange::Set { key, value } => {
                         TransactStateChange::Set { key, value }
                     }
                     transact::protocol::receipt::StateChange::Delete { key } => {
                         TransactStateChange::Delete { key }
                     }
                 })
+            }
+            TransactionResult::Invalid { .. } => panic!("already checked for invalid results"),
```

Comment:
The above could be reworked to avoid this panic (to avoid this match completely).

Reply:
Moved this conversion into Transact